x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -131,98 +124,95 @@ private void createFollowerIndex(
+                    // If restoreInfo is null then it is possible there was a master failure during the
+                    // restore.
+                    followingListener.onResponse(new PutFollowAction.Response(true, false, false));
+                } else if (restoreInfo.failedShards() == 0) {
+                    initiateFollowing(client, request, followingListener);
+                } else {
+                    // Has failed shards
+                    followingListener.onResponse(new PutFollowAction.Response(true, false, false));
                 }
-                imdBuilder.setRoutingNumShards(leaderIndexMetaData.getRoutingNumShards());
-                IndexMetaData followIMD = imdBuilder.build();
-                mdBuilder.put(followIMD, false);
+            }
 
-                ClusterState.Builder builder = ClusterState.builder(currentState);
-                builder.metaData(mdBuilder.build());
-                ClusterState updatedState = builder.build();
+            @Override
+            public void onFailure(Exception e) {
+                followingListener.onFailure(e);
+            }
+        };
+
+        Settings.Builder settingsBuilder = Settings.builder()
+            .put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getFollowRequest().getFollowerIndex())
+            .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+        String leaderClusterRepoName = CcrRepository.NAME_PREFIX + remoteCluster;
+        RestoreSnapshotRequest restoreRequest = new RestoreSnapshotRequest(leaderClusterRepoName, CcrRepository.LATEST)
+            .indices(request.getLeaderIndex()).indicesOptions(request.indicesOptions()).renamePattern("^(.*)$")
+            .renameReplacement(request.getFollowRequest().getFollowerIndex()).masterNodeTimeout(request.masterNodeTimeout())
+            .indexSettings(settingsBuilder);
+        initiateRestore(restoreRequest, restoreInitiatedListener, restoreCompleteHandler);
+    }
 
-                RoutingTable.Builder routingTableBuilder = RoutingTable.builder(updatedState.routingTable())
-                        .addAsNew(updatedState.metaData().index(request.getFollowRequest().getFollowerIndex()));
-                updatedState = allocationService.reroute(
-                        ClusterState.builder(updatedState).routingTable(routingTableBuilder.build()).build(),
-                        "follow index [" + request.getFollowRequest().getFollowerIndex() + "] created");
+    private void initiateRestore(RestoreSnapshotRequest restoreRequest, ActionListener<PutFollowAction.Response> restoreInitiatedListener,
+                                 ActionListener<RestoreSnapshotResponse> restoreCompleteListener) {
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new AbstractRunnable() {
+            @Override
+            public void onFailure(Exception e) {
+                restoreInitiatedListener.onFailure(e);
+                restoreCompleteListener.onFailure(e);
```

Comment:
I found the flow a bit hard to read with these two listeners. I've tried refactoring it as follows. Let me know whichever you prefer.

```
diff --git a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java
index 65c310859a9..1772054192e 100644
--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java
@@ -42,8 +42,6 @@ import java.util.Objects;
 public final class TransportPutFollowAction
     extends TransportMasterNodeAction<PutFollowAction.Request, PutFollowAction.Response> {
 
-    private static final ActionListener<PutFollowAction.Response> NOOP_LISTENER = ActionListener.wrap(() -> {});
-
     private final Client client;
     private final RestoreService restoreService;
     private final CcrLicenseChecker ccrLicenseChecker;
@@ -124,61 +122,18 @@ public final class TransportPutFollowAction
             return;
         }
 
-        String remoteCluster = request.getRemoteCluster();
-
-        Client client = CcrLicenseChecker.wrapClient(this.client, threadPool.getThreadContext().getHeaders());
-
-        final ActionListener<PutFollowAction.Response> followingListener;
-        final ActionListener<PutFollowAction.Response> restoreInitiatedListener;
-        if (request.getWaitForRestore()) {
-            followingListener = listener;
-            restoreInitiatedListener = NOOP_LISTENER;
-        } else {
-            followingListener = NOOP_LISTENER;
-            restoreInitiatedListener = listener;
-        }
-
-        ActionListener<RestoreSnapshotResponse> restoreCompleteHandler = new ActionListener<RestoreSnapshotResponse>() {
-            @Override
-            public void onResponse(RestoreSnapshotResponse restoreSnapshotResponse) {
-                RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
-
-                if (restoreInfo == null) {
-                    // If restoreInfo is null then it is possible there was a master failure during the
-                    // restore.
-                    followingListener.onResponse(new PutFollowAction.Response(true, false, false));
-                } else if (restoreInfo.failedShards() == 0) {
-                    initiateFollowing(client, request, followingListener);
-                } else {
-                    // Has failed shards
-                    followingListener.onResponse(new PutFollowAction.Response(true, false, false));
-                }
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                followingListener.onFailure(e);
-            }
-        };
-
-        Settings.Builder settingsBuilder = Settings.builder()
+        final Settings.Builder settingsBuilder = Settings.builder()
             .put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getFollowRequest().getFollowerIndex())
             .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
-        String leaderClusterRepoName = CcrRepository.NAME_PREFIX + remoteCluster;
-        RestoreSnapshotRequest restoreRequest = new RestoreSnapshotRequest(leaderClusterRepoName, CcrRepository.LATEST)
+        final String leaderClusterRepoName = CcrRepository.NAME_PREFIX + request.getRemoteCluster();
+        final RestoreSnapshotRequest restoreRequest = new RestoreSnapshotRequest(leaderClusterRepoName, CcrRepository.LATEST)
             .indices(request.getLeaderIndex()).indicesOptions(request.indicesOptions()).renamePattern("^(.*)$")
             .renameReplacement(request.getFollowRequest().getFollowerIndex()).masterNodeTimeout(request.masterNodeTimeout())
             .indexSettings(settingsBuilder);
-        initiateRestore(restoreRequest, restoreInitiatedListener, restoreCompleteHandler);
-    }
-
-    private void initiateRestore(RestoreSnapshotRequest restoreRequest, ActionListener<PutFollowAction.Response> restoreInitiatedListener,
-                                 ActionListener<RestoreSnapshotResponse> restoreCompleteListener) {
         threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new AbstractRunnable() {
             @Override
             public void onFailure(Exception e) {
-                restoreInitiatedListener.onFailure(e);
-                restoreCompleteListener.onFailure(e);
+                listener.onFailure(e);
             }
 
             @Override
@@ -186,14 +141,12 @@ public final class TransportPutFollowAction
                 restoreService.restoreSnapshot(restoreRequest, new ActionListener<RestoreService.RestoreCompletionResponse>() {
                     @Override
                     public void onResponse(RestoreService.RestoreCompletionResponse response) {
-                        restoreInitiatedListener.onResponse(new PutFollowAction.Response(true, false, false));
-                        RestoreClusterStateListener.createAndRegisterListener(clusterService, response, restoreCompleteListener);
+                        afterRestoreStarted(request, listener, response);
                     }
 
                     @Override
                     public void onFailure(Exception e) {
-                        restoreInitiatedListener.onFailure(e);
-                        restoreCompleteListener.onFailure(e);
+                        listener.onFailure(e);
                     }
                 });
             }
@@ -201,10 +154,57 @@ public final class TransportPutFollowAction
 
     }
 
+    private void afterRestoreStarted(
+        final PutFollowAction.Request request,
+        final ActionListener<PutFollowAction.Response> originalListener,
+        final RestoreService.RestoreCompletionResponse response) {
+
+        final ActionListener<PutFollowAction.Response> listener;
+        if (request.getWaitForRestore() == false) {
+            originalListener.onResponse(new PutFollowAction.Response(true, false, false));
+            listener = new ActionListener<PutFollowAction.Response>() {
+                @Override
+                public void onResponse(PutFollowAction.Response response) {
+                    logger.debug("put follow completed with {}", response); // TODO: implement toString() on Response object
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug("put follow failed after restore initiated", e);
+                }
+            };
+        } else {
+            listener = originalListener;
+        }
+
+        RestoreClusterStateListener.createAndRegisterListener(clusterService, response,
+            new ActionListener<RestoreSnapshotResponse>() {
+                @Override
+                public void onResponse(RestoreSnapshotResponse restoreSnapshotResponse) {
+                    final RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
+
+                    if (restoreInfo == null) {
+                        // If restoreInfo is null then it is possible there was a master failure during the restore.
+                        listener.onResponse(new PutFollowAction.Response(true, false, false));
+                    } else if (restoreInfo.failedShards() == 0) {
+                        initiateFollowing(request, listener);
+                    } else {
+                        // Has failed shards
+                        listener.onResponse(new PutFollowAction.Response(true, false, false));
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            });
+    }
+
     private void initiateFollowing(
-        final Client client,
         final PutFollowAction.Request request,
         final ActionListener<PutFollowAction.Response> listener) {
+        Client client = CcrLicenseChecker.wrapClient(this.client, threadPool.getThreadContext().getHeaders());
         client.execute(ResumeFollowAction.INSTANCE, request.getFollowRequest(), ActionListener.wrap(
             r -> listener.onResponse(new PutFollowAction.Response(true, true, r.isAcknowledged())),
             listener::onFailure
```

Reply:
I pulled in most of these changes.